File: zephyr/src/stream/stream.c
// Standard includes
#include <stdio.h>
#include <string.h>

// Zephyr includes
#include <zephyr/kernel.h>
#include <zephyr/logging/log.h>
#include <zephyr/net/socket.h>

// Cipher includes
#include <corekinect/cipher/protocol.h>
#include <corekinect/cipher/stream.h>
#include <corekinect/iface/iface.h>

#include "config/default.h"
#include "daemon/daemon.h"
#include "daemon/fifo.h"
#include "daemon/registry.h"
#include "utils/err.h"

// Private include
#include "interface.h"
#include "threads.h"

LOG_MODULE_REGISTER(stream, CONFIG_CK_CIPHER_STREAM_LOG_LEVEL);

/*-----------------------------------------------------------------------------------------------------
 *                                                                                 Private Data & Types
 *---------------------------------------------------------------------------------------------------*/

#define FNV1A_OFFSET_BASIS 2166136261u
#define FNV1A_PRIME        16777619u

// Single in-flight inbound stream (sufficient for point-to-point benchmarking).
static struct {
    bool in_progress;
    uint16_t stream_id;
    uint32_t total_len;
    uint32_t received_len;
    uint32_t num_chunks;
    uint32_t checksum;
    int64_t start_time;
} stream_rx;

static cipher_stream_rx_stats_t stream_last_rx;
static bool stream_last_rx_valid;
static uint32_t stream_completion_id;
static K_MUTEX_DEFINE(stream_mutex);

/*-----------------------------------------------------------------------------------------------------
 *                                                                                             Checksum
 *---------------------------------------------------------------------------------------------------*/
uint32_t cipher_stream_fnv1a(uint32_t seed, const uint8_t *data, size_t length) {
    uint32_t hash = seed;
    for (size_t i = 0; i < length; i++) {
        hash ^= data[i];
        hash *= FNV1A_PRIME;
    }
    return hash;
}

/*-----------------------------------------------------------------------------------------------------
 *                                                                                                 Send
 *---------------------------------------------------------------------------------------------------*/
static bool send_stream_packet(cipher_daemon_t *d, cipher_iface_t *iface, uint16_t dst,
                               uint16_t stream_id, uint8_t flag, uint16_t seq,
                               const uint8_t *payload, uint16_t payload_len) {
    cipher_packet_fifo_item_t *item = alloc_packet_fifo_item(d, payload_len);
    if (!item) {
        return false;
    }

    cipher_packet_t *packet = &item->packet;
    packet->header.source_id = d->device_id;
    packet->header.destination_id = dst;
    packet->header.service_id = stream_id;  // carries the stream id for correlation
    packet->header.operation_id = 0;
    packet->header.payload_len = payload_len;
    packet->header.sequence_num = seq;
    packet->header.type = CIPHER_PACKET_TYPE_STREAM;
    packet->header.hop_count = 0;
    packet->header.flags = flag;

    if (payload_len > 0) {
        memcpy(packet->payload, payload, payload_len);
    }

    // Back-pressured by the packet heap: alloc above blocks when the send
    // thread has not yet drained enough in-flight packets.
    k_fifo_put(&iface->decoded_packets_queue, item);
    return true;
}

int32_t cipher_stream_send(cipher_daemon_t *d, uint16_t device_id, uint16_t stream_id,
                           const uint8_t *data, uint32_t length, uint16_t chunk_size) {
    cipher_iface_t *iface = registry_get_iface(d, device_id);
    if (!iface) {
        LOG_ERR("stream: no route to device 0x%04x", device_id);
        return -1;
    }

    // Leave headroom so header + chunk stays within one framed packet.
    const uint16_t max_chunk = CONFIG_MAX_PAYLOAD_SIZE - sizeof(cipher_header_t) - 4;
    if (chunk_size == 0 || chunk_size > max_chunk) {
        chunk_size = max_chunk;
    }

    // START: total length + stream id
    uint8_t start_payload[6];
    memcpy(&start_payload[0], &length, sizeof(uint32_t));
    memcpy(&start_payload[4], &stream_id, sizeof(uint16_t));
    if (!send_stream_packet(d, iface, device_id, stream_id, CIPHER_FLAG_STREAM_START, 0,
                            start_payload, sizeof(start_payload))) {
        return -1;
    }

    // DATA: ordered chunks
    uint32_t sent = 0;
    uint16_t seq = 1;
    while (sent < length) {
        uint16_t n = (uint16_t)MIN((uint32_t)chunk_size, length - sent);
        if (!send_stream_packet(d, iface, device_id, stream_id, CIPHER_FLAG_STREAM_DATA, seq,
                                data + sent, n)) {
            return (int32_t)sent;
        }
        sent += n;
        seq++;
    }

    // END: declared length + checksum for verification
    uint32_t checksum = cipher_stream_fnv1a(FNV1A_OFFSET_BASIS, data, length);
    uint8_t end_payload[8];
    memcpy(&end_payload[0], &length, sizeof(uint32_t));
    memcpy(&end_payload[4], &checksum, sizeof(uint32_t));
    send_stream_packet(d, iface, device_id, stream_id, CIPHER_FLAG_STREAM_END, seq,
                       end_payload, sizeof(end_payload));

    return (int32_t)sent;
}

/*-----------------------------------------------------------------------------------------------------
 *                                                                                          Last Stream
 *---------------------------------------------------------------------------------------------------*/
bool cipher_stream_get_last_rx(cipher_daemon_t *d, cipher_stream_rx_stats_t *out) {
    ARG_UNUSED(d);
    bool valid;

    k_mutex_lock(&stream_mutex, K_FOREVER);
    valid = stream_last_rx_valid;
    if (valid) {
        *out = stream_last_rx;
    }
    k_mutex_unlock(&stream_mutex);

    return valid;
}

/*-----------------------------------------------------------------------------------------------------
 *                                                                                    Receive (Reassy)
 *---------------------------------------------------------------------------------------------------*/
static void handle_stream_packet(cipher_daemon_t *d, cipher_packet_t *packet) {
    ARG_UNUSED(d);
    const uint8_t *payload = (const uint8_t *)packet->payload;
    const uint16_t flags = packet->header.flags;

    if (flags & CIPHER_FLAG_STREAM_START) {
        k_mutex_lock(&stream_mutex, K_FOREVER);
        stream_rx.in_progress = true;
        memcpy(&stream_rx.total_len, &payload[0], sizeof(uint32_t));
        memcpy(&stream_rx.stream_id, &payload[4], sizeof(uint16_t));
        stream_rx.received_len = 0;
        stream_rx.num_chunks = 0;
        stream_rx.checksum = FNV1A_OFFSET_BASIS;
        stream_rx.start_time = k_uptime_get();
        k_mutex_unlock(&stream_mutex);
        LOG_INF("stream %u START: expecting %u bytes", stream_rx.stream_id, stream_rx.total_len);

    } else if (flags & CIPHER_FLAG_STREAM_DATA) {
        k_mutex_lock(&stream_mutex, K_FOREVER);
        stream_rx.received_len += packet->header.payload_len;
        stream_rx.num_chunks++;
        stream_rx.checksum = cipher_stream_fnv1a(stream_rx.checksum, payload, packet->header.payload_len);
        k_mutex_unlock(&stream_mutex);

    } else if (flags & CIPHER_FLAG_STREAM_END) {
        uint32_t declared_len = 0;
        uint32_t declared_checksum = 0;
        memcpy(&declared_len, &payload[0], sizeof(uint32_t));
        memcpy(&declared_checksum, &payload[4], sizeof(uint32_t));

        k_mutex_lock(&stream_mutex, K_FOREVER);
        int64_t duration = k_uptime_get() - stream_rx.start_time;
        stream_last_rx.completion_id = ++stream_completion_id;
        stream_last_rx.stream_id = stream_rx.stream_id;
        stream_last_rx.total_len = stream_rx.total_len;
        stream_last_rx.received_len = stream_rx.received_len;
        stream_last_rx.num_chunks = stream_rx.num_chunks;
        stream_last_rx.checksum = stream_rx.checksum;
        stream_last_rx.checksum_ok =
            (stream_rx.checksum == declared_checksum) && (stream_rx.received_len == declared_len);
        stream_last_rx.duration_ms = duration;
        stream_last_rx_valid = true;
        stream_rx.in_progress = false;
        cipher_stream_rx_stats_t snapshot = stream_last_rx;
        k_mutex_unlock(&stream_mutex);

        uint32_t kbps = (duration > 0) ? (uint32_t)((int64_t)snapshot.received_len * 1000 / duration / 1024) : 0;
        LOG_INF("stream %u END: %u bytes / %u chunks in %lld ms = %u KiB/s, checksum %s",
                snapshot.stream_id, snapshot.received_len, snapshot.num_chunks,
                duration, kbps, snapshot.checksum_ok ? "OK" : "MISMATCH");
    }
}

/*-----------------------------------------------------------------------------------------------------
 *                                                                                               Thread
 *---------------------------------------------------------------------------------------------------*/
void cipher_stream_thread(void *arg0, void *arg1, void *arg2) {
    ARG_UNUSED(arg1);
    ARG_UNUSED(arg2);
    cipher_daemon_t *d = (cipher_daemon_t *)arg0;

    while (true) {
        cipher_packet_fifo_item_t *fifo_item = k_fifo_get(&d->stream_packet_event_queue, K_FOREVER);
        __ASSERT(fifo_item, "Null item on stream_packet_event_queue, daemon %d", d->id);
        handle_stream_packet(d, &fifo_item->packet);
        free_packet_fifo_item(d, fifo_item);
    }
}

File: cipher/stream.go
package cipher

import (
	"encoding/binary"
	"fmt"
	"log"
	"time"

	"github.com/gophersys/cipher-go/iface"
)

/*-----------------------------------------------------------------------------------------------------
 *                                                                                             Checksum
 *---------------------------------------------------------------------------------------------------*/

const (
	fnv1aOffsetBasis uint32 = 2166136261
	fnv1aPrime       uint32 = 16777619
)

// FNV1a matches cipher_stream_fnv1a so a Go sender and a firmware receiver
// (and vice versa) agree on the stream checksum.
func FNV1a(seed uint32, data []byte) uint32 {
	hash := seed
	for _, b := range data {
		hash ^= uint32(b)
		hash *= fnv1aPrime
	}
	return hash
}

/*-----------------------------------------------------------------------------------------------------
 *                                                                                                 Send
 *---------------------------------------------------------------------------------------------------*/

// SendStream chunks data into ordered STREAM packets (START / DATA×N / END),
// wire-identical to cipher_stream_send. Control payloads are little-endian to
// match the firmware's raw-struct framing. Returns bytes sent.
func (d *Daemon) SendStream(deviceID, streamID uint16, data []byte, chunkSize int) (int, error) {
	transport := d.route(deviceID)
	if transport == nil {
		return 0, fmt.Errorf("cipher: no route to device 0x%04x", deviceID)
	}

	maxChunk := MaxPayloadSize - HeaderSize - 4
	if chunkSize <= 0 || chunkSize > maxChunk {
		chunkSize = maxChunk
	}

	// START: total length + stream id
	start := make([]byte, 6)
	binary.LittleEndian.PutUint32(start[0:4], uint32(len(data)))
	binary.LittleEndian.PutUint16(start[4:6], streamID)
	if err := d.sendStreamPacket(transport, deviceID, streamID, FlagStreamStart, 0, start); err != nil {
		return 0, err
	}

	// DATA: ordered chunks
	sent := 0
	seq := uint16(1)
	for sent < len(data) {
		n := chunkSize
		if remaining := len(data) - sent; remaining < n {
			n = remaining
		}
		if err := d.sendStreamPacket(transport, deviceID, streamID, FlagStreamData, seq, data[sent:sent+n]); err != nil {
			return sent, err
		}
		sent += n
		seq++
	}

	// END: declared length + checksum for verification
	end := make([]byte, 8)
	binary.LittleEndian.PutUint32(end[0:4], uint32(len(data)))
	binary.LittleEndian.PutUint32(end[4:8], FNV1a(fnv1aOffsetBasis, data))
	if err := d.sendStreamPacket(transport, deviceID, streamID, FlagStreamEnd, seq, end); err != nil {
		return sent, err
	}

	return sent, nil
}

func (d *Daemon) sendStreamPacket(transport *iface.Interface, dst, streamID uint16,
	flag uint8, seq uint16, payload []byte) error {
	header := Header{
		SourceID:      d.config.DeviceID,
		DestinationID: dst,
		ServiceID:     streamID, // carries the stream id for correlation
		Type:          PacketTypeStream,
		Flags:         flag,
		SequenceNum:   seq,
	}
	return d.sendPacket(transport, header, payload)
}

/*-----------------------------------------------------------------------------------------------------
 *                                                                                    Receive (Reassy)
 *---------------------------------------------------------------------------------------------------*/

// StreamStats mirrors cipher_stream_rx_stats_t.
type StreamStats struct {
	CompletionID uint32 // monotonic per receiving daemon, so readers can dedupe
	StreamID     uint16
	TotalLen     uint32
	ReceivedLen  uint32
	NumChunks    uint32
	Checksum     uint32
	ChecksumOK   bool
	DurationMs   int64
}

type streamReassembly struct {
	streamID  uint16
	totalLen  uint32
	received  uint32
	numChunks uint32
	checksum  uint32
	start     time.Time
}

// streamKey scopes a reassembly to the sending device AND the stream id.
// Keying by stream id alone let two peers streaming the same id clobber each
// other's in-flight reassembly.
func streamKey(sourceID, streamID uint16) uint32 {
	return uint32(sourceID)<<16 | uint32(streamID)
}

func (d *Daemon) handleStreamPacket(header Header, payload []byte) {
	d.streamMutex.Lock()
	defer d.streamMutex.Unlock()

	key := streamKey(header.SourceID, header.ServiceID)

	switch {
	case header.Flags&FlagStreamStart != 0:
		if len(payload) < 6 {
			return
		}
		d.streamRx[key] = &streamReassembly{
			streamID: binary.LittleEndian.Uint16(payload[4:6]),
			totalLen: binary.LittleEndian.Uint32(payload[0:4]),
			checksum: fnv1aOffsetBasis,
			start:    time.Now(),
		}

	case header.Flags&FlagStreamData != 0:
		if reassembly := d.streamRx[key]; reassembly != nil {
			reassembly.received += uint32(len(payload))
			reassembly.numChunks++
			reassembly.checksum = FNV1a(reassembly.checksum, payload)
		}

	case header.Flags&FlagStreamEnd != 0:
		reassembly := d.streamRx[key]
		if reassembly == nil || len(payload) < 8 {
			return
		}
		declaredLen := binary.LittleEndian.Uint32(payload[0:4])
		declaredChecksum := binary.LittleEndian.Uint32(payload[4:8])

		stats := StreamStats{
			StreamID:    reassembly.streamID,
			TotalLen:    reassembly.totalLen,
			ReceivedLen: reassembly.received,
			NumChunks:   reassembly.numChunks,
			Checksum:    reassembly.checksum,
			ChecksumOK:  reassembly.checksum == declaredChecksum && reassembly.received == declaredLen,
			DurationMs:  time.Since(reassembly.start).Milliseconds(),
		}
		d.streamCompletions++
		stats.CompletionID = d.streamCompletions
		d.lastStreamRx = stats
		d.lastStreamSet = true
		delete(d.streamRx, key)

		kibs := int64(0)
		if stats.DurationMs > 0 {
			kibs = int64(stats.ReceivedLen) * 1000 / stats.DurationMs / 1024
		}
		log.Printf("stream %d END: %d bytes / %d chunks in %d ms = %d KiB/s, checksum %v",
			stats.StreamID, stats.ReceivedLen, stats.NumChunks, stats.DurationMs, kibs, stats.ChecksumOK)
	}
}

// LastStreamRx returns stats for the most recently completed inbound stream.
func (d *Daemon) LastStreamRx() (StreamStats, bool) {
	d.streamMutex.Lock()
	defer d.streamMutex.Unlock()
	return d.lastStreamRx, d.lastStreamSet
}
